src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.c

```diff
@@ -0,0 +1,201 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "pal_eckey.h"
+
+#include <assert.h>
+
+EC_KEY* CryptoNative_NewEcKey(jobject curveParameters, jobject keyPair)
+{
+    assert(curveParameters);
+    assert(keyPair);
+
+    EC_KEY* keyInfo = malloc(sizeof(EC_KEY));
+    memset(keyInfo, 0, sizeof(EC_KEY));
+    keyInfo->refCount = 1;
+    keyInfo->curveParameters = curveParameters;
+    keyInfo->keyPair = keyPair;
+    return keyInfo;
+}
+
+void CryptoNative_EcKeyDestroy(EC_KEY* r)
+{
+    if (r)
+    {
+        r->refCount--;
+        if (r->refCount == 0)
+        {
+            JNIEnv* env = GetJNIEnv();
+            if (r->keyPair != NULL)
+            {
+                // Destroy the private key data.
+                jobject privateKey = (*env)->CallObjectMethod(env, r->keyPair, g_keyPairGetPrivateMethod);
+                if (privateKey)
+                {
+                    (*env)->CallVoidMethod(env, privateKey, g_destroy);
+                    ReleaseLRef(env, privateKey);
+                    CheckJNIExceptions(env); // The destroy call might throw an exception. Clear the exception state.
+                }
+            }
+
+            ReleaseGRef(env, r->keyPair);
+            ReleaseGRef(env, r->curveParameters);
+            free(r);
+        }
+    }
+}
+
+EC_KEY* CryptoNative_EcKeyCreateByOid(const char* oid)
+{
+    LOG_DEBUG("Creating curve from oid '%s'", oid);
+    JNIEnv* env = GetJNIEnv();
+
+    // Older versions of Android don't support mapping an OID to a curve name,
+    // so do some of the common mappings here.
+    jstring oidStr;
+    if (strcmp(oid, "1.3.132.0.33") == 0)
+    {
+        oidStr = JSTRING("secp224r1");
+    }
+    else if (strcmp(oid, "1.3.132.0.34") == 0 || strcmp(oid, "nistP384") == 0)
+    {
+        oidStr = JSTRING("secp384r1");
+    }
+    else if (strcmp(oid, "1.3.132.0.35") == 0 || strcmp(oid, "nistP521") == 0)
+    {
+        oidStr = JSTRING("secp521r1");
+    }
+    else if (strcmp(oid, "1.2.840.10045.3.1.7") == 0 || strcmp(oid, "nistP256") == 0)
+    {
+        oidStr = JSTRING("secp256r1");
+    }
+    else 
+    {
+        oidStr = JSTRING(oid);
+    }
+    jstring ec = JSTRING("EC");
+
+    // First, generate the key pair based on the curve defined by the oid.
+    jobject paramSpec = (*env)->NewObject(env, g_ECGenParameterSpecClass, g_ECGenParameterSpecCtor, oidStr);
+    ReleaseLRef(env, oidStr);
+
+    jobject keyPairGenerator =
```

Comment:
At least for me, my brain has already started re-parsing the JNI as though it's Java. I'm not the biggest fan of the comments since they're can easily get out of date and are sometimes just a simplification of what's happening (and not a direct 1:1).